### PR TITLE
Category protection

### PR DESCRIPTION
**Pull request by mic4ael**

### CODE REVIEW

File: indico/modules/categories/__init__.py [hunk 1]
```diff
@@ -86,6 +86,8 @@ def _sidemenu_items_old(sender, category, **kwargs):
 def _sidemenu_items(sender, category, **kwargs):
     yield SideMenuItem('content', _('Content'), url_for('categories.manage_content', category),
                        100, icon='eye')
+    yield SideMenuItem('protection', _('Protection'), url_for('categories.manage_protection', category), 70,
+                       icon='shield')
```
Comment (ThiefMaster) [inline indico/modules/categories/__init__.py]: I'd put the 70 in this line. More readable to have it consistent for all menu items

File: indico/modules/categories/forms.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+from indico.web.forms.widgets import SwitchWidget
+
+
+class CategoryProtectionForm(IndicoForm):
+    _event_creation_fields = ('event_creation_restricted', 'event_creators', 'event_creation_notification_emails')
+
+    protection_mode = IndicoProtectionField(_('Protection mode'), protected_object=lambda form: form.protected_object,
+                                            inherits_protection=lambda form: not form.protected_object.is_root)
+    acl = AccessControlListField(_('Access control list'),
+                                 groups=True, allow_emails=True, allow_external=True, allow_networks=True)
```
Comment (ThiefMaster) [inline indico/modules/categories/forms.py]: no emails allowed

File: indico/modules/categories/models/categories.py [hunk 3]
```diff
@@ -287,6 +287,9 @@ def get_chain_titles(self):
         """Retrieve the category titles from the root to the category"""
         return [c.title for c in self.chain_query]
 
+    def get_non_inheriting_objects(self):
```
Comment (ThiefMaster) [inline indico/modules/categories/models/categories.py]: why?

File: indico/web/forms/fields.py [hunk 4]
```diff
@@ -846,6 +846,9 @@ def __init__(self, *args, **kwargs):
         self.protected_object = kwargs.pop('protected_object')(kwargs['_form'])
         get_acl_message_url = kwargs.pop('acl_message_url', None)
         self.acl_message_url = get_acl_message_url(kwargs['_form']) if get_acl_message_url else None
+        self.inherits_protection = kwargs.pop('inherits_protection', lambda form: False)(kwargs['_form'])
+        if not self.inherits_protection:
+            kwargs['skip'] = [ProtectionMode.inheriting]
```
Comment (ThiefMaster) [inline indico/web/forms/fields.py]: `{ProtectionMode.inheriting}`

File: indico/modules/categories/models/categories.py (hunk 3 above)
Comment (mic4ael) [inline indico/modules/categories/models/categories.py]: Have a look at `IndicoProtectionField.render_protection_message` method. Either we override the method here or we put a proper `if` statement in the `render_protection_message`

File: indico/modules/categories/controllers/management.py [hunk 5]
```diff
@@ -16,10 +16,44 @@
...
+            return redirect(url_for('.manage_protection', self.category))
+        return WPCategoryManagement.render_template('management/category_protection.html', self.category, 'protection',
+                                                    form=form)
+
+    def _get_defaults(self):
+        acl = {x.principal for x in self.category.acl_entries if x.read_access}
+        managers = {x.principal for x in self.category.acl_entries if x.full_access}
+        event_creators = {x.principal for x in self.category.acl_entries
+                          if x.has_management_role('create', explicit=True)}
+        return {'protection_mode': self.category.protection_mode, 'acl': acl, 'managers': managers,
```
Comment (ThiefMaster) [inline indico/modules/categories/controllers/management.py]: Maybe see if `FormDefaults(self.category, acl=acl, managers=managers, ...)` works. Avoids repeating the "simple" fields. Probably you need to add acl, managers, etc to `skip_attrs` as FormDefaults prefers object attributes over kwargs.

File: indico/modules/categories/forms.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    acl = AccessControlListField(_('Access control list'), groups=True, allow_emails=False, allow_external=True,
+                                 allow_networks=True)
+    managers = PrincipalListField(_('Managers'), groups=True)
+    no_access_contact = StringField(_('No access contact'), description=_('Contact information shown when someone '
+                                                                          'lacks access to the category'))
+    event_creation_restricted = BooleanField(_('Restricted event creation'), widget=SwitchWidget(),
+                                             description=_('Whether the event creation should be restricted '
+                                                           'to a list of specific persons'))
+    event_creators = PrincipalListField(_('Event creators'), [HiddenUnless('event_creation_restricted')],
+                                        groups=True, allow_emails=True, allow_external=True,
```
Comment (ThiefMaster) [inline indico/modules/categories/forms.py]: no emails allowed

File: indico/web/forms/fields.py [hunk 7]
```diff
@@ -846,12 +846,18 @@ def __init__(self, *args, **kwargs):
         self.protected_object = kwargs.pop('protected_object')(kwargs['_form'])
         get_acl_message_url = kwargs.pop('acl_message_url', None)
         self.acl_message_url = get_acl_message_url(kwargs['_form']) if get_acl_message_url else None
+        self.inherits_protection = kwargs.pop('inherits_protection', lambda form: False)(kwargs['_form'])
```
Comment (ThiefMaster) [inline indico/web/forms/fields.py]: Not a fan of the naming. What you actually have here is something like `can_inherit_protection`

Comment (ThiefMaster) [inline indico/web/forms/fields.py]: Actually why not just check `protected_object.protection_parent is None`? Do we ever show this field in a form where `protected_object` is None? If not that'd be cleaner IMO.

File: indico/modules/categories/controllers/management.py [hunk 8]
```diff
@@ -16,10 +16,41 @@
...
+            return redirect(url_for('.manage_protection', self.category))
+        return WPCategoryManagement.render_template('management/category_protection.html', self.category, 'protection',
+                                                    form=form)
+
+    def _get_defaults(self):
+        acl = {x.principal for x in self.category.acl_entries if x.read_access}
+        managers = {x.principal for x in self.category.acl_entries if x.full_access}
+        event_creators = {x.principal for x in self.category.acl_entries
+                          if x.has_management_role('create', explicit=True)}
+        return {'acl': acl, 'managers': managers, 'event_creators': event_creators}
```
Comment (ThiefMaster) [inline indico/modules/categories/controllers/management.py]: why not return a `FormDefaults(self.category, acl=acl, ...)` object here?

File: indico/modules/categories/forms.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+from indico.web.forms.fields import AccessControlListField, PrincipalListField, IndicoProtectionField, EmailListField
+from indico.web.forms.validators import HiddenUnless
+from indico.web.forms.widgets import SwitchWidget
+
+
+class CategoryProtectionForm(IndicoForm):
+    _event_creation_fields = ('event_creation_restricted', 'event_creators', 'event_creation_notification_emails')
+
+    protection_mode = IndicoProtectionField(_('Protection mode'), protected_object=lambda form: form.protected_object)
+    acl = AccessControlListField(_('Access control list'), groups=True, allow_emails=False, allow_external=True,
```
Comment (ThiefMaster) [inline indico/modules/categories/forms.py]: `allow_emails=False` is the default, you could omit the kwarg

File: indico/modules/categories/forms.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+                                 allow_networks=True)
+    managers = PrincipalListField(_('Managers'), groups=True)
+    no_access_contact = StringField(_('No access contact'), description=_('Contact information shown when someone '
```
Comment (ThiefMaster) [inline indico/modules/categories/forms.py]: we usually put the `description` kwarg in a new line (-> less or no wrapping needed in the string in most cases)

**Pull request merged**